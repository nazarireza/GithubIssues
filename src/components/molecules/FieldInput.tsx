import React, { forwardRef, memo } from 'react';
import { TextInputProps, View, StyleSheet, TextInput } from 'react-native';
import { Input } from '../atoms/Input';
import { Label } from '../atoms/Label';
import { Space } from '../atoms/Space';

type FieldInputProps = {
  title?: string;
  hasError?: boolean;
  errorMessages?: string[];
};

export const FieldInput = memo(
  forwardRef<TextInput, FieldInputProps & TextInputProps>(
    ({ title, hasError, errorMessages, ...rest }, ref) => {
      return (
        <View>
          <Label>{title}</Label>
          <Space />
          <Input testID="input" ref={ref} {...rest} />
          <Space />
          {hasError && (
            <Label style={styles.errorMessage}>{errorMessages?.[0]}</Label>
          )}
        </View>
      );
    }
  )
);

const styles = StyleSheet.create({
  errorMessage: {
    color: 'red',
  },
});
