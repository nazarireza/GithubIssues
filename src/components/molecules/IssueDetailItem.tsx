import { memo } from 'react';
import { StyleSheet, View } from 'react-native';
import { IssueDto } from '../../services/types';
import { Label } from '../atoms/Label';
import { Space } from '../atoms/Space';
import Markdown from 'react-native-markdown-display';
import { IconButton } from '../atoms/IconButton';

type IssueDetailItemProps = {
  item: IssueDto;
  isBookmarked?: boolean;
  onToggleBookmark?: () => void;
};

export const IssueDetailItem: React.FC<IssueDetailItemProps> = memo(
  ({ item, isBookmarked = false, onToggleBookmark }) => {
    return (
      <View>
        <Label style={styles.title}>{item.title}</Label>
        <Space />
        <IconButton
          testID="bookmark"
          name={isBookmarked ? 'star' : 'star-outline'}
          onPress={onToggleBookmark}
        />
        <Space />
        <Markdown>{item.body}</Markdown>
      </View>
    );
  }
);

const styles = StyleSheet.create({
  title: {
    fontSize: 16,
    fontWeight: 'bold',
  },
  headerContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
});
