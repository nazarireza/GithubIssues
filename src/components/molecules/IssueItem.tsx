import { memo } from 'react';
import {
  StyleSheet,
  TouchableOpacity,
  TouchableOpacityProps,
} from 'react-native';
import { IssueDto } from '../../services/types';
import { Card } from '../atoms/Card';
import { Label } from '../atoms/Label';

type IssueItemProps = {
  item: IssueDto;
  isBookmarked?: boolean;
};

export const IssueItem: React.FC<IssueItemProps & TouchableOpacityProps> = memo(
  ({
    item: { id, title, body, state, created_at },
    isBookmarked = false,
    ...rest
  }) => {
    return (
      <TouchableOpacity activeOpacity={0.7} {...rest}>
        <Card>
          <Label style={styles.title}>{title}</Label>
        </Card>
      </TouchableOpacity>
    );
  }
);

const styles = StyleSheet.create({
  title: {
    fontSize: 16,
  },
});
