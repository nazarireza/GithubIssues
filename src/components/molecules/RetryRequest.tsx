import { memo } from 'react';
import { View, StyleSheet } from 'react-native';
import { useTranslation } from '../../assets/dictionary';
import { Button } from '../atoms/Button';
import { Label } from '../atoms/Label';
import { Space } from '../atoms/Space';

type RetryRequestProps = {
  message?: string;
  onRetry?: () => void;
};

export const RetryRequest: React.FC<RetryRequestProps> = memo(
  ({ message, onRetry }) => {
    const { t } = useTranslation();

    message = message || t("Something's Wrong");
    return (
      <View style={styles.container}>
        <Label>{message}</Label>
        <Space />
        <Button title={t('Retry')} onPress={onRetry} />
      </View>
    );
  }
);

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
  },
});
