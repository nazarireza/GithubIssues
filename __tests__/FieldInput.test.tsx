import React from 'react';
import renderer from 'react-test-renderer';
import { FieldInput } from '../src/components/molecules/FieldInput';
import { fireEvent, render } from 'react-native-testing-library'

describe('<FieldInput />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<FieldInput />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with title', () => {
    const tree = renderer.create(<FieldInput title="title1" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with error', () => {
    const tree = renderer
      .create(<FieldInput hasError errorMessages={['error1, error2']} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('fire the onChangeText', () => {
    const mockOnChangeText = jest.fn();
    const {getByTestId} = render(<FieldInput onChangeText={mockOnChangeText} />);

    const input = getByTestId('input');
    fireEvent(input, 'onChangeText', 'value1');

    expect(mockOnChangeText).toHaveBeenCalledWith('value1');
  })
});
