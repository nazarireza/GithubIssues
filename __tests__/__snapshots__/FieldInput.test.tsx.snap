// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldInput /> renders correctly 1`] = `
<View>
  <Text />
  <View
    style={
      Object {
        "height": 8,
        "width": 8,
      }
    }
  />
  <TextInput
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "paddingHorizontal": 8,
          "paddingVertical": 4,
        },
        undefined,
      ]
    }
    testID="input"
  />
  <View
    style={
      Object {
        "height": 8,
        "width": 8,
      }
    }
  />
</View>
`;

exports[`<FieldInput /> renders correctly with error 1`] = `
<View>
  <Text />
  <View
    style={
      Object {
        "height": 8,
        "width": 8,
      }
    }
  />
  <TextInput
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "paddingHorizontal": 8,
          "paddingVertical": 4,
        },
        undefined,
      ]
    }
    testID="input"
  />
  <View
    style={
      Object {
        "height": 8,
        "width": 8,
      }
    }
  />
  <Text
    style={
      Object {
        "color": "red",
      }
    }
  >
    error1, error2
  </Text>
</View>
`;

exports[`<FieldInput /> renders correctly with title 1`] = `
<View>
  <Text>
    title1
  </Text>
  <View
    style={
      Object {
        "height": 8,
        "width": 8,
      }
    }
  />
  <TextInput
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "paddingHorizontal": 8,
          "paddingVertical": 4,
        },
        undefined,
      ]
    }
    testID="input"
  />
  <View
    style={
      Object {
        "height": 8,
        "width": 8,
      }
    }
  />
</View>
`;
